File: Demo/VLFRunnerObjC/ViewController.m
//
//  ViewController.m
//  VLFRunnerObjC
//
//  Created by Sidharth Juyal on 08/07/2023.
//

#import "ViewController.h"
#import "VFLContext.h"
#import "VFLExample.h"
#import "VFLUtils.h"
#import "VFLViewController.h"

static NSString *kCellId = @"cell-id";

@interface ViewController () <UITableViewDataSource, UITableViewDelegate> {
  VFLContext *_context;
}
@property (nonatomic, readonly) VFLContext *vflContext;
@end

@implementation ViewController

- (void)viewDidLoad {
  [super viewDidLoad];
  self.title = @"VLF Demo";
  _context = [VFLContext context];
  
  [_context setParentView:self.view];
  UITableView *tableVw = [[UITableView alloc] initWithFrame:CGRectZero style:UITableViewStylePlain];
  [_context addSubviewsWithBindings:NSDictionaryOfVariableBindings(tableVw)];
  tableVw.dataSource = self;
  tableVw.delegate = self;
  [tableVw registerClass:[UITableViewCell class] forCellReuseIdentifier:kCellId];
}

- (void)viewSafeAreaInsetsDidChange {
  [super viewSafeAreaInsetsDidChange];
  [_context removeAllConstraints];
  [_context addMetrics:vfl_CreateUIEdgeInsetsMetrics(self.view.safeAreaInsets)];
  [_context addConstraintsWithFormats:
   [NSArray arrayWithObjects:
    @"H:|-(left)-[tableVw]-(right)-|",
    @"V:|-(top)-[tableVw]-(bottom)-|",
    nil]];
}

#pragma mark - UITableViewDataSource -
- (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section
{
  return VFLExampleTotal;
}

- (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath
{
  UITableViewCell *cell = [tableView dequeueReusableCellWithIdentifier:kCellId forIndexPath:indexPath];
  cell.textLabel.text = VFLExampleGetName(indexPath.row);
  cell.accessoryType = UITableViewCellAccessoryDisclosureIndicator;
  return cell;
}

- (void)tableView:(UITableView *)tableView didSelectRowAtIndexPath:(NSIndexPath *)indexPath {
  VFLViewController *vwCtrl = [VFLViewController viewControllerWithExample:indexPath.row];
  [self.navigationController pushViewController:vwCtrl animated:YES];
  [tableView deselectRowAtIndexPath:indexPath animated:YES];
}

@end

File: Demo/VFLRunner/Sources/ViewController.swift
//
//  ViewController.swift
//  VFLRunner
//
//  Created by Sidharth Juyal on 31/12/2022.
//

import UIKit
import VFL

class ViewController: UIViewController {
  static let cellId = "cell-id"
  
  lazy var vfl = VFL(view)
  
  override func viewDidLoad() {
    super.viewDidLoad()
    self.title = "VFL Demo"
    
    let tableVw = UITableView(frame: .zero, style: .plain)
    vfl
      .add(subview: tableVw, name: "tableVw")
    tableVw.dataSource = self
    tableVw.delegate = self
    tableVw.register(UITableViewCell.self, forCellReuseIdentifier: Self.cellId)
  }
  
  override func viewSafeAreaInsetsDidChange() {
    super.viewSafeAreaInsetsDidChange()
    vfl
      .replaceConstraints(
        metrics: view.safeAreaInsets.metrics,
        formats: [
          "V:|-(top)-[tableVw]-(bottom)-|",
          "H:|-(left)-[tableVw]-(right)-|"
        ])
  }
}

extension ViewController: UITableViewDataSource {
  func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
    return VFLExamples.allCases.count
  }
  
  func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
    let cell = tableView.dequeueReusableCell(withIdentifier: Self.cellId, for: indexPath)
    cell.textLabel?.text = VFLExamples.allCases[indexPath.row].rawValue
    cell.accessoryType = .disclosureIndicator
    return cell
  }
}

extension ViewController: UITableViewDelegate {
  func tableView(_ tableView: UITableView, didSelectRowAt indexPath: IndexPath) {
    let example = VFLExamples.allCases[indexPath.row]
    let vwCtrl = VFLViewController(contentView: example.view, title: example.rawValue)
    navigationController?.pushViewController(vwCtrl, animated: true)
    tableView.deselectRow(at: indexPath, animated: true)
  }
}
